8267265: Use new IR Test Framework to create tests for C2 IGV transformations

JohnTortugo opened the pull request:
Hi, can I please get some reviews for this Pull Request? Here is a summary of the changes:

- Add tests, using the new IR-based test framework, for several of the Ideal transformations on Add, Sub, Mul, Div, Loop nodes and some simple Scalar Replacement transformations.
- Add more default IR regex's to IR-based test framework.
- Changes to Sub, Div and Add Ideal nodes to that transformations on Int and Long types are the whenever possible same.
- Changes to Sub*Node, Div*Node and Add*Node Ideal methods to fix some bugs and include new transformations.
- New JTREG "ir_transformations" test group under test/hotspot/jtreg.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8267265](https://bugs.openjdk.java.net/browse/JDK-8267265): Use new IR Test Framework to create tests for C2 IGV transformations


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5135/head:pull/5135` \
`$ git checkout pull/5135`

Update a local copy of the PR: \
`$ git checkout pull/5135` \
`$ git pull https://git.openjdk.java.net/jdk pull/5135/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5135`

View PR using the GUI difftool: \
`$ git pr show -t 5135`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5135.diff">https://git.openjdk.java.net/jdk/pull/5135.diff</a>

</details>

JohnTortugo commented:
/contributor @jtfuller111

JohnTortugo commented:
/contributor add @jtfuller111

JohnTortugo commented:
@chhagedorn - I finally managed to get some time to work again on this. Can you please take a look?

chhagedorn reviewed (changes requested):
First of all, thanks a lot for diving into the various C2 IR transformations and creating IR tests for them! It's valuable to have a test for each of them. Some general points:

- I think you should split the work into the creation of new IR tests of existing transformations (this RFE) and changing actual C2 IR transformations. I suggest to file a new RFE or bug for missing/potential wrong transformations with the corresponding IR tests for them.
- I have only skimmed through the IR tests, yet. I will wait with reviewing them closer in case you decide to transform some of them into custom run tests (see comments in files).
- You should add the `hotspot-compiler` label to the PR as the tests are for C2.

Thanks,
Christian

iignatev commented on test/hotspot/jtreg/TEST.groups:
```diff
@@ -71,6 +71,9 @@ hotspot_native_sanity = \
 
+hotspot_ir_transformations = \
+  ir_transformations
+
```
as these tests are c2 tests, they are to be inside `test/hotspot/jtreg/compiler/c2`. I don't think `ir_framework` is a good directory name, as they aren't tests for `ir_framework`, nor `ir_framework` itself,  so I'd suggest `test/hotspot/jtreg/compiler/c2/ir_transformations`, or `test/hotspot/jtreg/compiler/c2/idealization`.

on a separate note, @JohnTortugo why have you added a dedicated group for these tests?

-- Igor

chhagedorn replied:
Thanks Igor for your input!

> as they aren't tests for `ir_framework`

Right. This might be misleading. 

>  so I'd suggest test/hotspot/jtreg/compiler/c2/ir_transformations, or test/hotspot/jtreg/compiler/c2/idealization.

Could we also choose `test/hotspot/jtreg/compiler/c2/ir_tests` (should we use camel case or underscore?) as top folder for any kinds of IR tests? Then we could still add a subfolder like `ideal_transformations` to it (`test/hotspot/jtreg/compiler/c2/ir_tests/ideal_transformations`). If there are other IR tests later, for example for string opts, we can add a `stringopts` subfolder into the `ir_tests` top folder and so on.

iignatev replied:
> Could we also choose `test/hotspot/jtreg/compiler/c2/ir_tests` as top folder for any kinds of IR tests? 

sounds good to me. 

> should we use camel case or underscore?

given it's java code, and the directory names are part of a package name, we should use lowerCamelCase.

vnkozlov replied:
I am **against** moving tests under `/c2` directory. This directory was created to split original set of tests to run in parallel and not related to C2 compiler. I want to have directories in `/compiler` per optimization/transformation type as we have now. For example, `/loopopts` is for C2 only but it is not under `/c2`.

Actually I want to clear `c2` and `c1` directories eventually to move tests to corresponding category directories instead.

So I am for `/compiler/irTests`.

iignatev replied:
> This directory was created to split original set of tests to run in parallel and not related to C2 compiler

aren't you mixing this up w/ `tier1_compiler_[1-3]` test group? if you aren't, this was the wrong way to split tests and we need to fix that.

>  I want to have directories in `/compiler` per `optimization/transformation` type as we have now. For example, `/loopopts` is for C2 only but it is not under `/c2`.
> Actually I want to clear `c2` and `c1` directories eventually to move tests to corresponding category directories instead.

I would prefer to clear `/compiler` directory and move c1-specific tests into `/compiler/c1` directory and c2-specific tests into `/compiler/c2` with optimization/transformation type being subdirectories inside `/compiler/c[12]` (and shared/common tests moved to `/compiler/shared`), so `loopopts` tests are to moved to `/compiler/c2/loopopts`. one of the reasons, most of optimizations tests will be compiler-specific and having them together in `/compiler` doesn't make much sense and just makes it harder to navigate/use test base. e.g. if we have all c2 tests in `/compiler/c2`, one can easily run them together to test c2 changes, if c2-specific tests are in several directories in `/compiler`, so one will have to have intrinsic knowledge about all tests in `/compiler` to select the right one.

JohnTortugo replied:
Should I move all  tests created in this PR to folders under `/compiler/c2/`? Separated by type of optimization - loopOpts, scalarReplacement, etc?

JohnTortugo replied:
I moved the tests to `/compiler/c2/irTests/`. Please let me know if the split into subfolders I did is reasonable.

JohnTortugo commented on test/hotspot/jtreg/ir_transformations/AddINodeIdealizationTests.java:
```diff
@@ -0,0 +1,187 @@
+    }
+
+    @Test
+    @Arguments(Argument.RANDOM_EACH)
+    @IR(failOn = {IRNode.LOAD, IRNode.STORE, IRNode.MUL, IRNode.DIV, IRNode.SUB})
+    // Checks (x - x) => 0 and 0 - 0 => 0
+    public int simpleZeroSub(int x) {
+        return (x - x) + (x - x);
+    }
```
Thank you @chhagedorn , I think this is a good idea. I'll follow your suggestion and transform some tests into `custom run tests`.

chhagedorn replied:
Great, thanks! Btw, you can merge and now use `RunInfo.getRandom().XX()` for a handy access to random values (if needed) as the PR for JDK-8272567 was integrated in the meantime.

JohnTortugo replied:
Hi, again @chhagedorn. I added some `custom run tests` to tests that seemed more "complex". Please let me know if there are others that you think I should add.

TobiHartmann replied:
@JohnTortugo just FYI, @chhagedorn is currently on vacation but will be back mid next week.

JohnTortugo replied:
Thank you @TobiHartmann

JohnTortugo commented on test/hotspot/jtreg/compiler/lib/ir_framework/IRNode.java:
```diff
@@ -45,45 +45,50 @@
  * @see IR
  */
 public class IRNode {
```
Sounds good to me!

chhagedorn replied:
Thanks!

JohnTortugo commented on test/hotspot/jtreg/ir_transformations/AddINodeIdealizationTests.java:
```diff
@@ -0,0 +1,187 @@
+        TestFramework.run();
+    }
+
+    @Test
+    @Arguments(Argument.RANDOM_EACH)
+    @IR(failOn = {IRNode.LOAD, IRNode.STORE, IRNode.MUL, IRNode.DIV, IRNode.SUB})
+    @IR(counts = {IRNode.ADD, "2"})
+    // Checks
+    public int simpleZero(int x) {
+        return (x + x) + (x + x);
+    }
```
I think the "Zero" suffix was actually an index, as in "simple-0", "simple-1", etc. Clearly not a good name in this case. I'll rename it. Thanks.

chhagedorn replied:
Ah I see, it was a little bit confusing :-)

JohnTortugo commented on src/hotspot/share/opto/divnode.cpp:
```diff
@@ -466,6 +466,20 @@ Node *DivINode::Ideal(PhaseGVN *phase, bool can_reshape) {
+  Node *in2 = in(2);
+  uint op1 = in1->Opcode();
+
+  // Converts (x * y) / y => x
+  if (op1 == Op_MulI && in1->in(2) == in2) {
+    return new AddINode(in1->in(1), phase->intcon(0));
+  }
+
+  // Converts (x * y) / x => y
+  if (op1 == Op_MulI && in1->in(1) == in2) {
+    return new AddINode(in1->in(2), phase->intcon(0));
+  }
```
Oops! Thanks for catching that. I'll remove it and the test.

chhagedorn replied:
Sounds good, thanks!

JohnTortugo commented:
/label add hotspot-compiler

JohnTortugo commented:
>     * I think you should split the work into the creation of new IR tests of existing transformations (this RFE) and changing actual C2 IR transformations. I suggest to file a new RFE or bug for missing/potential wrong transformations with the corresponding IR tests for them.

I just created [this RFE](https://bugs.openjdk.java.net/browse/JDK-8272735) for the discussion/work on adjusting the Ideal transformations.

JohnTortugo commented:
Hi folks,

I'd appreciate if you could take a look again. I addressed the comments so far and here is a summary of the latest changes:

- Moved all tests to directory `/compiler/c2/irTests` under hotspot JTREG group.
- Added a few `custom run tests` for testing some corner cases.
- Renamed some tests files and some test methods.
- Added new tests for testing the changes introduced by [JDK-8270823](https://bugs.openjdk.java.net/browse/JDK-8270823)

chhagedorn commented on test/hotspot/jtreg/compiler/lib/ir_framework/IRNode.java:
```diff
@@ -129,6 +160,23 @@
     public static final String SCOPE_OBJECT = "(.*# ScObj.*" + END;
     public static final String MEMBAR = START + "MemBar" + MID + END;
 
```
I suggest to move all newly added regex together here.

chhagedorn reviewed (changes requested):
Thanks for your effort to write tests for all these different kinds of transformations! Generally, they look good and are worth to have! You should add `@bug 8267265` to all files.

JohnTortugo commented on test/hotspot/jtreg/compiler/c2/irTests/AddINodeIdealizationTests.java:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    @Test
+    @IR(failOn = {IRNode.LOAD, IRNode.STORE, IRNode.MUL, IRNode.DIV, IRNode.SUB})
```
Do you think I can remove the "LOAD" and "STORE" as well?

chhagedorn replied:
Yes, I think you can remove them, too, as long as there are no fields involved.

JohnTortugo commented on test/hotspot/jtreg/compiler/c2/irTests/MulINodeIdealizationTests.java:
```diff
@@ -0,0 +1,157 @@
+    @IR(counts = {IRNode.MUL, "1"})
+    //Checks Max(a,b) * min(a,b) => a*b
+    public int excludeMaxMin(int x, int y){
+        return Math.max(x, y) * Math.min(x, y);
```
Created this work item to address the suggestion: https://bugs.openjdk.java.net/browse/JDK-8274799
